test/integration/ssl_integration_test.cc

```diff
@@ -121,31 +99,68 @@ TEST_P(SslIntegrationTest, RouterRequestAndResponseWithBodyNoBufferVerifySAN) {
 TEST_P(SslIntegrationTest, RouterRequestAndResponseWithBodyNoBufferHttp2VerifySAN) {
   setDownstreamProtocol(Http::CodecClient::Type::HTTP2);
   ConnectionCreationFunction creator = [&]() -> Network::ClientConnectionPtr {
-    return makeSslClientConnection(true, true);
+    return makeSslClientConnection(ClientSslTransportOptions().setAlpn(true).setSan(true));
   };
   testRouterRequestAndResponseWithBody(1024, 512, false, &creator);
   checkStats();
 }
 
+// Server with an ECDSA certificate and a client with RSA/ECDSA cipher suites
+// works.
+TEST_P(SslIntegrationTest, RouterRequestAndResponseWithBodyNoBufferServerEcdsa) {
+  server_ecdsa_cert_ = true;
+  ConnectionCreationFunction creator = [&]() -> Network::ClientConnectionPtr {
+    return makeSslClientConnection({});
+  };
+  testRouterRequestAndResponseWithBody(1024, 512, false, &creator);
+  checkStats();
+}
+
+// Server with an RSA certificate and a client with only RSA cipher suites
+// works.
+TEST_P(SslIntegrationTest, RouterRequestAndResponseWithBodyNoBufferClientRsaOnly) {
+  ConnectionCreationFunction creator = [&]() -> Network::ClientConnectionPtr {
+    return makeSslClientConnection(
+        ClientSslTransportOptions().setCipherSuites({"ECDHE-RSA-AES128-GCM-SHA256"}));
+  };
+  testRouterRequestAndResponseWithBody(1024, 512, false, &creator);
+  checkStats();
+}
+
+// Server has only an ECDSA certificate, client is only RSA capable, leads
+// to a connection fail.
+TEST_P(SslIntegrationTest, RouterRequestAndResponseWithBodyNoBufferServerEcdsaClientRsaOnly) {
+  server_ecdsa_cert_ = true;
+  initialize();
+  EXPECT_FALSE(
+      makeRawHttpConnection(makeSslClientConnection(ClientSslTransportOptions().setCipherSuites(
+                                {"ECDHE-RSA-AES128-GCM-SHA256"})))
+          ->connected());
+  Stats::CounterSharedPtr counter =
+      test_server_->counter(listenerStatPrefix("ssl.connection_error"));
+  EXPECT_EQ(1U, counter->value());
+  counter->reset();
+}
+
```

Comment:
You're missing the case with RSA server and ECDSA client.

Reply:
This is currently implied, but sure, I can make an explicit variant.